### DISCUSSION TITLE
Production Deployment Package

### DISCUSSION THREAD

**Discussion by betimer** [vercel/next.js#12135] [Help]
So can I confirm for production deployment package, what it only needs is:

`.next` folder that has been built
`node_modules` packages

It does not require the source code. Is this correct?

**Comment by timneutkens** [comment]
Correct! the source code is not needed, only node_modules and .next

**Reply by Janpot** [reply]
Is `node_modules` required with the `serverless` target?

**Reply by timneutkens** [reply]
@Janpot depends on which serverless target you use, the `serverless` target does inline all deps and there it's not needed, however for the serverless target you shouldn't upload the full `.next` folder either, it runs in separate parts.

**Reply by NMinhNguyen** [reply]
> however for the serverless target you shouldn't upload the full `.next` folder either, it runs in separate parts.

@timneutkens could you clarify what's necessary? Is it just the `serverless` folder? From what I understand, `serverless` target with `next start` and `.next` (or maybe just `.next/serverless`) should be sufficient to deploy? Although I would still need `next` with all of its dependencies available presumably?

https://nextjs.org/docs/api-reference/next.config.js/build-target#serverless-target

> It's only compatible with `next start` or Serverless deployment platforms (like Vercel) — you cannot use the custom server API.

The reason I'm asking for clarification is I saw https://github.com/vercel/next.js/issues/6173#issuecomment-583736184 and although I don't have an immediate need for independent serverless functions as such, I'd like the ability to not have to deploy a big `node_modules` folder. The quoted text below makes me think that `next start` with `serverless` should handle serving static pages as well as executing lambdas.

> However a Next.js app since Next.js 9 is no longer "just a server" as said in my earlier reply. You can't get away with only using AWS Lambda, you also have to serve static pages. When #9524 lands it'll become even more complicated as you'll have to know how to route a page to a static file _and_ lambdas for the same route.

**Reply by timneutkens** [reply]
> serverless target with next start

Your application will be less performant if you use the serverless target and then use `next start` to boot up as a lot of modules won't be shared between page rendering, causing higher cpu and memory load. The serverless target is meant to output multiple serverless functions and is not meant to be deployed as a server. If you're deploying to a server it's better to use `next build` and then `next start` where node_modules and .next are available on your server.

**Reply by wei** [reply]
@timneutkens thanks for your detailed response. For small sites like the one I'm working on, `node_modules` excluding `devDependencies` is already at 150MB-250MB range while `.next/` folder with `serverless` build is only a dozen MB with `experimental-serverless-trace` it's only 4MB. This results in quite a difference in the docker image being built. Is it still recommended to choose `server` target over `serverless` despite the large difference in docker image size caused by `node_modules`?

You mentioned above

> depends on which serverless target you use, the serverless target does inline all deps and there it's not needed

and before in [10741](https://github.com/vercel/next.js/discussions/10741#discussioncomment-745)
> Later on we added another target called experimental-serverless-trace, this target is much more optimized version of serverless that works more like how server would currently, except it's better than server as it no longer requires a full node_modules to be shipped.

after I ran `next build` with `serverless` target and moved `.next` to a new folder, I ran `npm install next` then `next start`. It's throwing module not found errors for `react`, `react-dom`, etc. Do I still have to install those even though all the dependencies are supposed to already been packed in various js files inside `.next/serverless` folder? Is it a limitation of using `next start` with `serverless` target? I tried the same steps with `experimental-serverless-trace` target and had the same error.

Lastly, have you seen any other lighter server compatibility layer alternatives to using `next start` with the `serverless` output files? I ask because `next` is quite heavy with tons of dependencies in `node_modules`. If none is out there atm, have you guys thought about making the next server into its own package?

Thanks so much in advance!
